w11b01/test/pgdp/pengusurvivors/ResidualGraphTest.java

```diff
@@ -0,0 +1,178 @@
+package pgdp.pengusurvivors;
+
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import java.util.List;
+import java.util.stream.Stream;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class ResidualGraphTest {
+
+
+    @ParameterizedTest
+    @MethodSource("getResidualGraphArguments")
+    void shouldGenerateCorrectResidualGraph(String source, String expectedResidual) {
+        var sourceGraph = GraphGenerator.generateGraph(source).graph();
+        var residualGraph = GraphGenerator.generateGraph(expectedResidual).graph();
+        sourceGraph.generateResidualGraph();
+        assertTrue(GraphUtils.residualEquals(sourceGraph, residualGraph));
+    }
+
+    static Stream<Arguments> getResidualGraphArguments() {
+        return Stream.of(
+                Arguments.of(
+                        """
+                                a -> b (4/10);
+                                b -> c (2/6);
+                                """,
+                        """
+                                a -> b (0/6);
+                                b -> a (0/4);
+                                b -> c (0/4);
+                                c -> b (0/2);
+                                """
+                )
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("findPathArguments")
+    void findPath(String verticeString, String source, String sink, List<List<String>> expectedPaths) {
+        var graphResult = GraphGenerator.generateGraph(
+                verticeString,
+                source,
+                sink
+        );
+        var graph = graphResult.graph();
+        GraphUtils.printGraph(graph);
+        var vertices = graphResult.vertices();
+        graph.generateResidualGraph();
+        var path = graph.findPathInResidual();
+        var matchingPaths = 0;
```

Comment:
It would be nice to always output the generated solution.

```suggestion
        var path = graph.findPathInResidual();

        System.out.println("Your path was: " + (path == null ? "None found" : GraphUtils.stringifyPath(path)));

        var matchingPaths = 0;
```